File: stock_search/news/providers/massive.py
"""Massive
Documentation: https://massive.com/docs/rest/stocks/news
- Free
- 5 API Calls / Minute
"""

import asyncio
from datetime import UTC, datetime, timedelta
import os

from dotenv import load_dotenv
import httpx

from ...models.schemas import NewsArticle, NewsMetadata
from ...utils import extract_domain, format_date, get_days_ago, parse_date

load_dotenv()

SENTIMENT_MAP = {
    "positive": "bullish",
    "neutral": "neutral",
    "negative": "bearish",
}
DEFAULT_TIMEOUT_SEC = 60


async def get_news_massive_async(
    ticker: str,
    n_days: int = 3,
    max_results: int = 25,
    *,
    client: httpx.AsyncClient,
) -> list[NewsArticle]:
    """Get financial news using Massive API."""
    params = {
        "apiKey": os.getenv("MASSIVE_API_KEY"),
        "ticker": ticker,
        "published_utc.gte": (datetime.now(UTC) - timedelta(days=n_days)).isoformat(),
        "order": "desc",
        "limit": max_results,
        "sort": "published_utc",
    }
    response = await client.get(
        url="https://api.massive.com/v2/reference/news",
        params=params,
    )
    response.raise_for_status()
    results = response.json().get("results", [])

    news_list = []
    fetched_at = datetime.now(UTC).isoformat()
    for result in results:
        dt = parse_date(result["published_utc"])
        insights = result.get("insights", [{}])
        sentiment = SENTIMENT_MAP.get(insights[0].get("sentiment"), "neutral")
        url = result["article_url"]
        news_list.append(
            NewsArticle(
                title=result["title"],
                url=url,
                date=format_date(dt),
                days_ago=get_days_ago(dt),
                summary=result.get("description", "[FAILED TO FETCH]"),
                sentiment=sentiment,
                metadata=NewsMetadata(
                    provider="massive",
                    source_domain=extract_domain(url),
                    published_at=dt.astimezone(UTC).isoformat(),
                    fetched_at=fetched_at,
                ),
            )
        )
    return news_list


def get_news_massive(
    ticker: str,
    n_days: int = 3,
    max_results: int = 25,
) -> list[NewsArticle]:
    """Get financial news using Massive API."""

    async def _fetch() -> list[NewsArticle]:
        async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT_SEC) as client:
            return await get_news_massive_async(
                ticker=ticker,
                n_days=n_days,
                max_results=max_results,
                client=client,
            )

    return asyncio.run(_fetch())

File: src/stock-search/news/providers/massive.ts
/** Massive
Documentation: https://massive.com/docs/rest/stocks/news
- Free
- 5 API Calls / Minute
- Max 1000 results per query
*/

import {
	daysAgo,
	fetchJson,
	formatDate,
	normalizeDomain,
	parseDate,
	readJsonResponse,
	DAY_IN_MS,
} from "./shared.js";
import { newsArticleSchema, type NewsArticle } from "../../models/schemas.js";

const SENTIMENT_MAP: Record<string, NewsArticle["sentiment"]> = {
	positive: "bullish",
	neutral: "neutral",
	negative: "bearish",
};

export const MASSIVE_MAX_RESULTS = 1000;

export async function getNewsMassiveAsync({
	ticker,
	nDays = 3,
	maxResults = 25,
	client,
}: {
	ticker: string;
	nDays?: number;
	maxResults?: number;
	client?: {
		get: (input: {
			url: string;
			params: Record<string, string>;
		}) => Promise<{ json(): Promise<unknown> | unknown; raise_for_status?: () => void }>;
	};
}): Promise<NewsArticle[]> {
	const boundedMaxResults = Math.min(maxResults, MASSIVE_MAX_RESULTS);
	const params = {
		apiKey: process.env.MASSIVE_API_KEY ?? "",
		ticker,
		"published_utc.gte": new Date(Date.now() - nDays * DAY_IN_MS).toISOString(),
		order: "desc",
		limit: String(boundedMaxResults),
		sort: "published_utc",
	};
	const url = "https://api.massive.com/v2/reference/news";
	const payload = client
		? await readJsonResponse<{ results?: Array<Record<string, unknown>> } | null>(
				await client.get({ url, params }),
			)
		: await fetchJson<{ results?: Array<Record<string, unknown>> }>(
				`${url}?${new URLSearchParams(params).toString()}`,
			);

	const fetchedAt = new Date().toISOString();
	return (payload?.results ?? []).map((row) => {
		const publishedAt = parseDate(row.published_utc);
		const insights = Array.isArray(row.insights) ? row.insights : [{}];
		const sentiment =
			SENTIMENT_MAP[String((insights[0] as Record<string, unknown>).sentiment ?? "")] ??
			"neutral";
		const rawUrl = String(row.article_url ?? "");
		return newsArticleSchema.parse({
			url: rawUrl,
			title: String(row.title ?? rawUrl),
			date: formatDate(publishedAt),
			days_ago: daysAgo(publishedAt),
			summary: String(row.description ?? "[FAILED TO FETCH]"),
			sentiment,
			metadata: {
				provider: "massive",
				source_domain: normalizeDomain(rawUrl),
				published_at: publishedAt?.toISOString() ?? null,
				fetched_at: fetchedAt,
			},
		});
	});
}
